### PR TITLE
Attachments must be in json format

### PR DESCRIPTION
**Pull request by agentsib**
Uncorrect attachment serialization. Slack not accept attachment without encoding.

### CODE REVIEW

**Comment by agentsib** [comment]
Need think how to check json string in test...

**Comment by agentsib** [comment]
Test "ChatPostMessagePayloadTest" where dirty. I think need costomize it. 

And maybe need add json serialzation in method "addAttachement"? @cleentfaar, that you thinks?

**Comment by cleentfaar** [comment]
Hi @agentsib thanks for reporting this! 

There may be an issue with Attachments not being sent properly that I will have to look into; can you give me a *working* example of using attachments with the API (so just the raw post URL to the slack API)? I am asking this because I have not even been able to sent attachments normally, so this is hard to test for me.

On your suggestion for adding it to `addAttachment`; imo encoding should not be done in the model itself because of separation of concerns. 
Think of the implications: every time you call `getAttachments` you would have to decode the stored JSON back to PHP; this is not expected behaviour for a model imo, at least not in the way I intended them for this library.

Aside from that I don't think this is an encoding issue, but perhaps you can give me a scenario that I can reproduce it with?

**Comment by cleentfaar** [comment]
Sorry @agentsib, seems my last comment was a bit too soon; this is a nice solution! Does it (sending attachments) work for you with this code?

I only have a few comments about the code itself, see line-notes below

File: src/CL/Slack/Payload/AdvanceSerializeInterface.php [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+
+namespace CL\Slack\Payload;
+
+
+use JMS\Serializer\Serializer;
+
+interface AdvanceSerializeInterface
```
Comment (cleentfaar) [inline src/CL/Slack/Payload/AdvanceSerializeInterface.php]: Can you refactor this to `AdvancedSerializerInterface` (including the extra `d` and `r`)

File: src/CL/Slack/Serializer/PayloadSerializer.php [hunk 2]
```diff
@@ -25,6 +27,9 @@ class PayloadSerializer extends AbstractSerializer
      */
     public function serialize(PayloadInterface $payload)
     {
+        if ($payload instanceof AdvanceSerializeInterface) {
+            $payload->beforeSerialize($this->serializer);
+        }
```
Comment (cleentfaar) [inline src/CL/Slack/Serializer/PayloadSerializer.php]: Can you add a newline below this for readability?

**Comment by agentsib** [comment]
I fix check style. https://github.com/agentsib/slacktest - this is a demo, please try this :) Thanks)

**Comment by bramdevries** [comment]
Any update on when this can be merged?

**Comment by bramdevries** [comment]
@cleentfaar any update?

**Comment by Stayallive** [comment]
@cleentfaar any updates on this? This is quite frustrating that there is no response and that this is an issue since may :)

If you are not satisfied with the PR please tell us what needs to be done!

**Comment by cleentfaar** [comment]
Hey guys I really appreciate your contributions and am sorry for keeping you waiting so long!

I will be making some time this weekend to backtrack everything that have been submitted the last months. 

Work and (other) hobbies have taken over much of my time recently, hope you understand!

**Comment by cleentfaar** [comment]
Thanks for this PR @agentsib ! Again sorry to keep you waiting

**Comment by Stayallive** [comment]
Great, thanks @cleentfaar! And thanks @agentsib for making this PR <3

**Pull request merged**